### ISSUE TITLE
Dataflow: Static field used in inner class

### ISSUE THREAD

**Issue by smillst** [typetools/checker-framework#818]
A static field of can be referred to with or without the classname: `field` or `MyClass.field`.  Dataflow does not recognize these as the same field in an inner class.  For example: 

````
import org.checkerframework.checker.nullness.qual.*;
class StaticField {
    public static @Nullable Object o = null;
    class StaticInnerClass {
        void method() {
           StaticField.o = new Object();
           o.toString(); // false positive dereference.of.nullable
        }
        void method2() {
           o = new Object();
           StaticField.o.toString();// false positive dereference.of.nullable
        }
    }
}
````
More test cases in checker/tests/nullness/flow/Issue818.java

Blocking #764

**Comment by stefanheule** [comment]
I've been asked if I have any ideas what's wrong here, so here are my thoughts after investigating a little:

I'm relatively sure the error happens during construction of the CFG in CFGBuilder.  Specifically, the else branch in the [getReceiver](https://github.com/typetools/checker-framework/blob/8c77df601e95a84ce4bec80a65e480a477abfea8/dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java#L2603).  In the else branch we handle the case where a field access "f" is done without a receiver, and so we create an ImplicitReceiverNode.  This is correct, except that the type we give to that receiver is wrong.  We just pick the enclosing class, yet this only works if the field is in fact declared on the enclosing class.  As this issue shows, this does not work for nested classes.

Fixing this should be relatively easy:  One would have to check which class the field was declared on, and the give the right type to the ImplicitReceiverNode.

Hope this helps.

**Comment by smillst** [comment]
Thanks for looking into this!

In the case of a static field or method, should created receiver be a ClassNameNode rather than a ImplicitThisLiteralNode?  (Or a maybe a new ImplicitClassNameNode?)

**Comment by stefanheule** [comment]
That is a good question, I guess so far we use an ImplicitThisLiteralNode even for static fields (both in the case where the field is on the current class, or an enclosing class like in the test case).  It seems that something like ImplicitClassNameNode would be more accurate, but I don't know if this is necessary.  Without trying anything, my guess is that ImplicitClassNameNode is more accurate, but takes more work to implement.

**Comment by smillst** [comment]
Ok.  Thanks for the insight!
